File: python_test_crate/tests/13_system_return.py
from typing import Any, Optional, List, Dict, Callable

class SystemReturnTest:
    def __init__(self):
        self._state_stack = []
        self._state_context = {}
        self._return_value = None
        self._state = "Calculator"
        self._enter()

    def _transition(self, target_state, exit_args = None, enter_args = None):
        if exit_args:
            self._exit(*exit_args)
        else:
            self._exit()
        self._state = target_state
        if enter_args:
            self._enter(*enter_args)
        else:
            self._enter()

    def _change_state(self, target_state):
        self._state = target_state

    def _dispatch_event(self, event, *args):
        handler_name = f"_s_{self._state}_{event}"
        handler = getattr(self, handler_name, None)
        if handler:
            return handler(*args)

    def _enter(self, *args):
        if self._state == "Calculator":
            self._state_context["value"] = 0


    def _exit(self, *args):
        # No exit handlers
        pass

    def add(self, a: int, b: int) -> int:
        self._return_value = None
        self._dispatch_event("add", a, b)
        return self._return_value

    def multiply(self, a: int, b: int) -> int:
        self._return_value = None
        self._dispatch_event("multiply", a, b)
        return self._return_value

    def greet(self, name: str) -> str:
        self._return_value = None
        self._dispatch_event("greet", name)
        return self._return_value

    def get_value(self) -> int:
        self._return_value = None
        self._dispatch_event("get_value")
        return self._return_value

    def _s_Calculator_greet(self, name: str) -> str:
        self._return_value = "Hello, " + name + "!"
        return

    def _s_Calculator_multiply(self, a: int, b: int) -> int:
        self._return_value = a * b

    def _s_Calculator_get_value(self) -> int:
        self._state_context["value"] = 42
        self._return_value = self._state_context["value"]
        return

    def _s_Calculator_add(self, a: int, b: int) -> int:
        self._return_value = a + b
        return


def main():
    print("=== Test 13: System Return ===")
    calc = SystemReturnTest()

    # Test return sugar
    result = calc.add(3, 5)
    assert result == 8, f"Expected 8, got {result}"
    print(f"add(3, 5) = {result}")

    # Test system.return = expr
    result = calc.multiply(4, 7)
    assert result == 28, f"Expected 28, got {result}"
    print(f"multiply(4, 7) = {result}")

    # Test string return
    greeting = calc.greet("World")
    assert greeting == "Hello, World!", f"Expected 'Hello, World!', got '{greeting}'"
    print(f"greet('World') = {greeting}")

    # Test return with state variable
    value = calc.get_value()
    assert value == 42, f"Expected 42, got {value}"
    print(f"get_value() = {value}")

    print("PASS: System return works correctly")

if __name__ == '__main__':
    main()

File: rust_test_crate/tests/13_system_return.rs
use std::collections::HashMap;

#[derive(Clone, Default)]
struct CalculatorContext {
    value: i32,
}

#[derive(Clone)]
enum SystemReturnTestCompartment {
    Calculator(CalculatorContext),
    Empty,
}

impl Default for SystemReturnTestCompartment {
    fn default() -> Self {
        SystemReturnTestCompartment::Calculator(CalculatorContext::default())
    }
}

pub struct SystemReturnTest {
    _state: String,
    _state_stack: Vec<(String, SystemReturnTestCompartment)>,
    _state_context: HashMap<String, Box<dyn std::any::Any>>,
    _sv_value: i32,
}

impl SystemReturnTest {
    pub fn new() -> Self {
        let mut this = Self {
            _state_stack: vec![],
            _state_context: HashMap::from([]),
            _sv_value: 0,
            _state: String::from("Calculator"),
        };
        this._enter();
        this
    }

    fn _transition(&mut self, target_state: &str) {
        self._exit();
        self._state = target_state.to_string();
        self._enter();
    }

    fn _dispatch_event(&mut self, event: &str) {
let handler_name = format!("_s_{}_{}", self._state, event);
// Rust requires match-based dispatch or a handler registry
// For now, use explicit match in caller;
    }

    fn _enter(&mut self) {
match self._state.as_str() {
    "Calculator" => {
        self._sv_value = 0;
    }
    _ => {}
}
    }

    fn _exit(&mut self) {
        // No exit handlers
    }

    fn _state_stack_push(&mut self) {
let compartment = match self._state.as_str() {
    "Calculator" => SystemReturnTestCompartment::Calculator(CalculatorContext { value: self._sv_value }),
    _ => SystemReturnTestCompartment::Empty,
};
self._state_stack.push((self._state.clone(), compartment));
    }

    fn _state_stack_pop(&mut self) {
let (saved_state, compartment) = self._state_stack.pop().unwrap();
self._exit();
self._state = saved_state;
match compartment {
    SystemReturnTestCompartment::Calculator(ctx) => {
        self._sv_value = ctx.value;
    }
    SystemReturnTestCompartment::Empty => {}
}
    }

    pub fn add(&mut self, a: i32, b: i32) -> i32 {
match self._state.as_str() {
            "Calculator" => self._s_Calculator_add(a, b),
            _ => Default::default(),
        }
    }

    pub fn multiply(&mut self, a: i32, b: i32) -> i32 {
match self._state.as_str() {
            "Calculator" => self._s_Calculator_multiply(a, b),
            _ => Default::default(),
        }
    }

    pub fn get_value(&mut self) -> i32 {
match self._state.as_str() {
            "Calculator" => self._s_Calculator_get_value(),
            _ => Default::default(),
        }
    }

    fn _s_Calculator_get_value(&mut self) -> i32 {
self._sv_value = 42;
return self._sv_value
    }

    fn _s_Calculator_add(&mut self, a: i32, b: i32) -> i32 {
return a + b
    }

    fn _s_Calculator_multiply(&mut self, a: i32, b: i32) -> i32 {
return a * b
    }
}


fn main() {
    println!("=== Test 13: System Return ===");
    let mut calc = SystemReturnTest::new();

    // Test caret return sugar
    let result = calc.add(3, 5);
    assert_eq!(result, 8, "Expected 8, got {}", result);
    println!("add(3, 5) = {}", result);

    // Test system.return = expr
    let result = calc.multiply(4, 7);
    assert_eq!(result, 28, "Expected 28, got {}", result);
    println!("multiply(4, 7) = {}", result);

    // Test return with state variable
    let value = calc.get_value();
    assert_eq!(value, 42, "Expected 42, got {}", value);
    println!("get_value() = {}", value);

    println!("PASS: System return works correctly");
}
